src/stage1/broadcast.jl

```diff
@@ -29,44 +29,54 @@ function (∂ₙ::∂☆{N})(zc::ZeroBundle{N, typeof(copy)},
   return r
 end
 
-_print(s) = nothing
-# _print(s) = printstyled(s, "\n"; color=:magenta)
+# Reverse mode broadcast rules
+
+using ChainRulesCore: derivatives_given_output
+
+# _print(s) = nothing
+_print(s) = printstyled(s, "\n"; color=:magenta)
 
 # Broadcast over one element is just map
 function (∂⃖ₙ::∂⃖{N})(::typeof(broadcasted), f, a::Array) where {N}
     _print("path 0")
     ∂⃖ₙ(map, f, a)
 end
 
-using ChainRulesCore: derivatives_given_output
-
 (::∂⃖{1})(::typeof(broadcasted), f, args...) = split_bc_rule(f, args...)
 (::∂⃖{1})(::typeof(broadcasted), f, arg::Array) = split_bc_rule(f, arg) # ambiguity
 function split_bc_rule(f::F, args...) where {F}
     T = Broadcast.combine_eltypes(f, args)
-    if T == Bool
-        # Trivial case
+    TΔ = Core.Compiler._return_type(derivatives_given_output, Tuple{T, F, map(eltype, args)...})
+    if eltype(T) == Bool
+        # Trivial case: non-differentiable output
         _print("path 1")
         back_1(_) = ntuple(Returns(ZeroTangent()), length(args)+2)
         return f.(args...), back_1
-    elseif isconcretetype(Core.Compiler._return_type(
-            derivatives_given_output, Tuple{T, F, map(eltype, args)...}))
+    elseif T <: Number && isconcretetype(TΔ)
         # Fast path: just broadcast, and use x & y to find derivative.
         ys = f.(args...)
         _print("path 2")
-        function back_2(dys)
+        function back_2_one(dys)  # For f.(x) we do not need StructArrays / unzip at all
+            delta = broadcast(unthunk(dys), ys, args...) do dy, y, a
+                das = only(derivatives_given_output(y, f, a))
+                dy * conj(only(das))
+            end
+            (NoTangent(), NoTangent(), unbroadcast(only(args), delta))
+        end
+        function back_2_many(dys)
             deltas = splitcast(unthunk(dys), ys, args...) do dy, y, as...
                 das = only(derivatives_given_output(y, f, as...))
                 map(da -> dy * conj(da), das)
             end
-            dargs = map(unbroadcast, args, deltas)
+            dargs = map(unbroadcast, args, deltas)  # ideally sum in unbroadcast could be part of splitcast?
             (NoTangent(), NoTangent(), dargs...)
         end
-        return ys, back_2
+        return ys, length(args)==1 ? back_2_one : back_2_many
```

Comment:
Indeed, good point. The length can be made a type parameter `Vararg{N}` though so it ought to be known in advance? The other choices depend on inference, less sure.

Reply:
It's not type stable, although it seems to be the type of the primal which is Any, not the function. I think from the 3rd case, where it's the first part of `∂⃖{1}().(f, args...)`, instead of simply broadcasting. That's also the case which prevents moving the branches inside the pullback.